src/thor/triplegbuilder.cc

```diff
@@ -35,32 +35,38 @@ void TrimShape(std::vector<PointLL>& shape,
                const PointLL& start_vertex,
                const float end,
                const PointLL& end_vertex) {
-  // clip up to the start point
+  // clip up to the start point if the start_vertex is valid
   float along = 0.f;
   auto current = shape.begin();
-  while (current != shape.end() - 1) {
-    along += (current + 1)->Distance(*current);
-    // just crossed it
-    if ((along > start) && start_vertex.IsValid()) {
-      along = start;
-      *current = start_vertex;
-      shape.erase(shape.begin(), current);
-      break;
+  if (start_vertex.IsValid()) {
+    while (current != shape.end() - 1) {
```

Comment:
why not add the condition to the while loop to reduce the indentation and the diff?

Reply:
start_vertex never changes inside the loop so why recheck it over the entire shape?

Reply:
I could add this after the first loop: 
  // return if end vertex is not valid
  if (!end_vertex.IsValid()) {
    return;
  }

that would reduce some diffs

Reply:
because branch prediction makes it pretty much free and its easier to read. was just a nit pick nothing too pressing

Reply:
this logic (and similar logic for "partial" shape) has always been tricky. I wanted to add some comments around it and make it clearer.